platform/linuxbsd/display_server_x11.cpp

```diff
@@ -2325,6 +2345,56 @@ void DisplayServerX11::_send_window_event(const WindowData &wd, WindowEvent p_ev
 	}
 }
 
+void DisplayServerX11::_poll_events_thread(void *ud) {
+	DisplayServerX11 *display_server = (DisplayServerX11 *)ud;
+	display_server->_poll_events();
+}
+
+void DisplayServerX11::_poll_events() {
+	int x11_fd = ConnectionNumber(x11_display);
+	fd_set in_fds;
+
+	while (!events_thread_done) {
+		XFlush(x11_display);
+
+		FD_ZERO(&in_fds);
+		FD_SET(x11_fd, &in_fds);
+
+		struct timeval tv;
+		tv.tv_usec = 0;
+		tv.tv_sec = 1;
+
+		// Wait for next event or timeout.
+		int num_ready_fds = select(x11_fd + 1, &in_fds, NULL, NULL, &tv);
+		if (num_ready_fds < 0) {
+			ERR_PRINT("_poll_events: select error: " + itos(errno));
+		}
+
+		// Process events from the queue.
+		{
+			MutexLock mutex_lock(events_mutex);
+
+			auto predicate = [](Display *display, XEvent *event, XPointer arg) -> Bool {
+				// Just accept all events.
+				return True;
+			};
+
+			// Non-blocking wait for next event
+			// and remove it from the queue.
+			XEvent ev;
+			while (XCheckIfEvent(x11_display, &ev, predicate, nullptr)) {
+				// Check if the input manager wants to process the event.
+				if (XFilterEvent(&ev, None)) {
+					// Event has been filtered by the Input Manager,
+					// it has to be ignored and a new one will be received.
+					continue;
+				}
+				polled_events.push_back(ev);
+			}
+		}
+	}
+}
```

Comment:
Do you not need some sort of locking to protect from calls to x11_display in the main thread? like for setting window size and other stuff

Reply:
XLib calls generally use XLockDisplay/XUnlockDisplay internally, so it should be safe as long as they don't poll events from the event queue while our thread process is doing it as well. I've used our own mutex when this case happens.

Reply:
Maybe @hpvb can validate that I'm not overlooking anything about thread safety.